### PR TITLE
Fix old calendar dav urls

### LINKED ISSUE #21812: [CalDAV/CardDAV] core's dav app has to provide old caldav and carddav endpoints for already setup clients
Clients which are already connected to an ownCloud instance use the webdav endpoints
- remote.php/caldav
- remote.php/carddav

We need to implement these endpoints within the dav app of core in oc9.0 as well

### LINKED ISSUE #22346: Shared calendar urls not working anymore after update
### Steps
1. Install 8.2.2
* Create a calendar test as user1
* Share calendar with user2
* In your caldav client of choice add: `remote.php/caldav/calendars/user2/test_shared_by_user1`
* Update to 9.0-beta1

### Expected
You can still access and sync the calendar

### Actual
404 - Not Found

### Info
`remote.php/dav/calendars/user2/test_shared_by_user1` (dav instead of caldav as first folder) works, but we want a smooth transition, so the old links should still work.

### PR DESCRIPTION
**Pull request by nickvergessen**
See https://github.com/owncloud/core/issues/22346 for steps

Fix #22346
Fix #21812 

@DeepDiver1975 @PVince81

### CODE REVIEW

File: apps/dav/appinfo/v1/caldav.php [hunk 1]
```diff
@@ -29,26 +29,32 @@
...
+	\OC::$server->getGroupManager()
 );
 $db = \OC::$server->getDatabaseConnection();
-$calDavBackend = new CalDavBackend($db, $principalBackend);
+$calDavBackend = new CalDavBackend($db, $userPrincipalBackend);
 
 // Root nodes
-$principalCollection = new \Sabre\CalDAV\Principal\Collection($principalBackend);
-$principalCollection->disableListing = true; // Disable listing
+$userPrincipals = new \Sabre\CalDAV\Principal\Collection($userPrincipalBackend, 'principals/users');
```
Comment (DeepDiver1975) [inline apps/dav/appinfo/v1/caldav.php]: this is wrong - the principal url for the old caldav and carddav url only contains users and doesnot contain the users subfolder

**Comment by nickvergessen** [comment]
Okay, now the own calendars work again.

However shared calendars and contacts fail with:
```xml
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:sabredav-version>3.0.7</s:sabredav-version>
  <s:exception>Sabre\DAVACL\Exception\NeedPrivileges</s:exception>
  <s:message>User did not have the required privileges ({DAV:}read) for path "calendars/user1/test_shared_by_user2"</s:message>
  <d:need-privileges>
    <d:resource>
      <d:href>/remote.php/caldav/calendars/user1/test_shared_by_user2</d:href>
      <d:privilege>
        <d:read/>
      </d:privilege>
    </d:resource>
  </d:need-privileges>
</d:error>
```

The problem is when checking privileges, we check if `'principal/user1'` is in `['principal/users/user1']`

Continuing

**Comment by nickvergessen** [comment]
possible fix would be to overwrite the DAVACL plugin:
```php
class AuthPlugin extends \Sabre\DAVACL\Plugin {

	/**
	 * Returns the standard users' principal.
	 *
	 * This is one authorative principal url for the current user.
	 * This method will return null if the user wasn't logged in.
	 *
	 * @return string|null
	 */
	function getCurrentUserPrincipal() {
		$principal = parent::getCurrentUserPrincipal();
		return $this->convertPrincipal($principal);
	}


	/**
	 * Returns a list of principals that's associated to the current
	 * user, either directly or through group membership.
	 *
	 * @return array
	 */
	function getCurrentUserPrincipals() {
		$currentUser = $this->getCurrentUserPrincipal();

		if (is_null($currentUser)) return [];

		return [$currentUser];
	}

	private function convertPrincipal($principal) {
		if (strpos($principal, 'principals/users/') !== 0) {
			list(, $name) = URLUtil::splitPath($principal);
			return "principals/users/$name";
		}
		return $principal;
	}
}
```

But not sure if that is the correct way @DeepDiver1975

**Comment by DeepDiver1975** [comment]
bloody mess

I'll have a look on MOnday - THX

**Comment by karlitschek** [comment]
cool. thanks

**Comment by DeepDiver1975** [comment]
> possible fix would be to overwrite the DAVACL plugin:

@nickvergessen looks like a reasonable approach - wana push it?

File: apps/dav/lib/connector/legacydavacl.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\DAV\Connector;
+
+
+use Sabre\HTTP\URLUtil;
+
+class LegacyDAVACL extends \Sabre\DAVACL\Plugin {
```
Comment (nickvergessen) [inline apps/dav/lib/connector/legacydavacl.php]: Not sure, if this is the best class name, but it's only needed for V1, and I thought `V1DAVACL` looks even more odd.

**Comment by nickvergessen** [comment]
Updated @DeepDiver1975

**Comment by DeepDiver1975** [comment]
tested and works - THX a lot! :+1:

**Comment by nickvergessen** [comment]
@schiesbn willing to do a local update test of your instance, to confirm this?

**Comment by DeepDiver1975** [comment]
@schiesbn mind testing? THX

**Comment by nickvergessen** [comment]
or maybe @georgehrke or @PVince81 are willing to test?

**Comment by rullzer** [comment]
Works here :+1: 
This will make my family very happy :) Thanks.

**Comment by DeepDiver1975** [comment]
need rebase ....

**Pull request merged**